### DISCUSSION TITLE
Default to single threaded actions

### DISCUSSION THREAD

**Discussion by blackliner** [bazelbuild/rules_rust#3101] [Ideas]
We have a decently sized rust codebase (800k LOC), and we regularly suffer from the fact that rust actions spawn up to N parallel threads. It leads to CPU starvation (slow, but not ciritcal) and OOM events (very bad!!).

I think there are two approaches
1. make bazel aware of the "size" of an action, see https://github.com/bazelbuild/bazel/issues/10443. Likely a bigger effort, but eventually the right approach IMHO
2. default to single threaded executions for compilation (codegen-units) and tests (test-threads)

I see the second approach being [used in this repo](https://github.com/search?q=repo%3Abazelbuild%2Frules_rust+codegen-units&type=code), but from the surrounding comments I deduce this is done because of better optimized build results.

**References:**
- Bazel high level issue: https://github.com/bazelbuild/bazel/issues/10443
- Other languages have similar issues, see for example the [bazel + Nix rules](https://github.com/tweag/rules_nixpkgs/issues/482), where the number of parallel jobs can go N² (blowing up your memory)

**Comment by UebelAndre** [comment]
This might be relevant to https://github.com/bazelbuild/rules_rust/issues/3008

**Reply by UebelAndre** [reply]
I'd be happy to review a change to add representative resource tracking to `Rustc` actions. I just haven't had the time to track down if actions by default are trying to take 16 CPU. If so, we should make a global flag in Bazel that explicitly controls that and then ensure all actions use that.

**Reply by blackliner** [reply]
with https://github.com/bazelbuild/rules_rust/pull/3112 merged, is this problem "solved"?

**Comment by hofbi** [comment]
@UebelAndre @illicitonion  I think there is a similar issue for tests. According to https://doc.rust-lang.org/rustc/tests/index.html#--test-threads-num_threads

> By default, uses the amount of concurrency available on the hardware as indicated by [available_parallelism](https://doc.rust-lang.org/std/thread/fn.available_parallelism.html).

So I wonder if it would make sense to set `--test-threads=1` to make sure bazel is in control of all parallel actions.
